From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Franz-Josef Haider <franz.haider@jolla.com>
Date: Thu, 8 Oct 2020 14:43:42 +0000
Subject: [PATCH] (hybris) Fix __get_tls and related functions (>=Android 10)

Adapted for Android 16: __get_tls changed from macros to static inline
functions. We replace them with extern declarations so libhybris can
intercept TLS access. __set_tls functions are preserved as-is.
Added riscv support in tls_internal.h (new in A16).

Signed-off-by: Simonas Leleiva <simonas.leleiva@jolla.com>
Change-Id: I24486873be638a1a8cd123705ba2e51ec7414b94
---
 libc/Android.bp                         |  1 +
 libc/bionic/ndk_cruft.cpp               | 12 ++++---
 libc/bionic/pthread_internal.h          |  6 ++--
 libc/hybris_support.c                   |  4 +++
 libc/libc.map.txt                       |  1 +
 libc/platform/bionic/tls.h              | 41 ++++++----------------
 libc/platform/bionic/tls_internal.h     | 46 +++++++++++++++++++++++++
 libc/platform/scudo_platform_tls_slot.h |  2 +-
 8 files changed, 75 insertions(+), 38 deletions(-)
 create mode 100644 libc/hybris_support.c
 create mode 100644 libc/platform/bionic/tls_internal.h

diff --git a/libc/Android.bp b/libc/Android.bp
index efba83b6e..acea5c9a1 100644
--- a/libc/Android.bp
+++ b/libc/Android.bp
@@ -975,6 +975,7 @@ cc_library_static {
         "stdio/stdio_ext.cpp",
         "stdio/vfscanf.cpp",
         "stdio/vfwscanf.cpp",
+        "hybris_support.c",
     ],
 
     arch: {
diff --git a/libc/bionic/ndk_cruft.cpp b/libc/bionic/ndk_cruft.cpp
index bc06d9dc6..0689000e5 100644
--- a/libc/bionic/ndk_cruft.cpp
+++ b/libc/bionic/ndk_cruft.cpp
@@ -66,6 +66,13 @@
 
 extern "C" {
 
+// TODO: does anything still need this?
+// with hybris we need it, so thanks for keeping it.
+void** __get_tls() {
+#include "platform/bionic/tls_internal.h"
+    return __get_tls_internal();
+}
+
 // By the time any NDK-built code is running, there are plenty of threads.
 int __isthreaded = 1;
 
@@ -85,10 +92,7 @@ int __open() {
   abort();
 }
 
-// TODO: does anything still need this?
-void** __get_tls() {
-  return __real_get_tls();
-}
+
 
 // This non-standard function was in our <string.h> for some reason.
 void memswap(void* m1, void* m2, size_t n) {
diff --git a/libc/bionic/pthread_internal.h b/libc/bionic/pthread_internal.h
index b3133a92d..722c761a6 100644
--- a/libc/bionic/pthread_internal.h
+++ b/libc/bionic/pthread_internal.h
@@ -223,16 +223,16 @@ __LIBC_HIDDEN__ void* __allocate_stack_mte_ringbuffer(size_t n, pthread_internal
 #endif
 
 static inline __always_inline bionic_tcb* __get_bionic_tcb() {
-  return reinterpret_cast<bionic_tcb*>(&__get_tls()[MIN_TLS_SLOT]);
+  return reinterpret_cast<bionic_tcb*>(&__get_tls_hooks()[MIN_TLS_SLOT]);
 }
 
 // Make __get_thread() inlined for performance reason. See http://b/19825434.
 static inline __always_inline pthread_internal_t* __get_thread() {
-  return static_cast<pthread_internal_t*>(__get_tls()[TLS_SLOT_THREAD_ID]);
+  return static_cast<pthread_internal_t*>(__get_tls_hooks()[TLS_SLOT_THREAD_ID]);
 }
 
 static inline __always_inline bionic_tls& __get_bionic_tls() {
-  return *static_cast<bionic_tls*>(__get_tls()[TLS_SLOT_BIONIC_TLS]);
+  return *static_cast<bionic_tls*>(__get_tls_hooks()[TLS_SLOT_BIONIC_TLS]);
 }
 
 static inline __always_inline TlsDtv* __get_tcb_dtv(bionic_tcb* tcb) {
diff --git a/libc/hybris_support.c b/libc/hybris_support.c
new file mode 100644
index 000000000..2784c85ec
--- /dev/null
+++ b/libc/hybris_support.c
@@ -0,0 +1,4 @@
+void** __get_tls_hooks(void) {
+#include "platform/bionic/tls_internal.h"
+  return __get_tls_internal();
+}
diff --git a/libc/libc.map.txt b/libc/libc.map.txt
index 3e9f8505f..d5bfad052 100644
--- a/libc/libc.map.txt
+++ b/libc/libc.map.txt
@@ -40,6 +40,7 @@ LIBC {
     __fsetlocking; # introduced=23
     __fstatfs64; # arm x86
     __fwritable; # introduced=23
+    __get_tls_hooks; # arm arm64 x86 introduced=21
     __get_h_errno;
     __getcpu; # arm x86
     __getcwd; # arm x86
diff --git a/libc/platform/bionic/tls.h b/libc/platform/bionic/tls.h
index 3e0ef53c6..59e078ff8 100644
--- a/libc/platform/bionic/tls.h
+++ b/libc/platform/bionic/tls.h
@@ -30,13 +30,16 @@
 
 #include <sys/cdefs.h>
 
-#if defined(__aarch64__)
-
-static inline void** __get_tls(void) {
-  void** result;
-  __asm__("mrs %0, tpidr_el0" : "=r"(result));
-  return result;
+#ifdef __cplusplus
+extern "C" {
+#endif
+void** __get_tls(void);
+void** __get_tls_hooks(void);
+#ifdef __cplusplus
 }
+#endif
+
+#if defined(__aarch64__)
 
 static inline void __set_tls(void* tls) {
   __asm__("msr tpidr_el0, %0" : : "r" (tls));
@@ -44,12 +47,6 @@ static inline void __set_tls(void* tls) {
 
 #elif defined(__arm__)
 
-static inline void** __get_tls(void) {
-  void** result;
-  __asm__("mrc p15, 0, %0, c13, c0, 3" : "=r"(result));
-  return result;
-}
-
 // arm32 requires a syscall to set the thread pointer.
 // By historical accident it's public API, but not in any header except this one.
 __BEGIN_DECLS
@@ -58,12 +55,6 @@ __END_DECLS
 
 #elif defined(__i386__)
 
-static inline void** __get_tls(void) {
-  void** result;
-  __asm__("movl %%gs:0, %0" : "=r"(result));
-  return result;
-}
-
 // x86 is really hairy, so we keep that out of line.
 __BEGIN_DECLS
 int __set_tls(void* tls);
@@ -71,24 +62,12 @@ __END_DECLS
 
 #elif defined(__riscv)
 
-static inline void** __get_tls(void) {
-  void** result;
-  __asm__("mv %0, tp" : "=r"(result));
-  return result;
-}
-
 static inline void __set_tls(void* tls) {
   __asm__("mv tp, %0" : : "r"(tls));
 }
 
 #elif defined(__x86_64__)
 
-static inline void** __get_tls(void) {
-  void** result;
-  __asm__("mov %%fs:0, %0" : "=r"(result));
-  return result;
-}
-
 // ARCH_SET_FS is not exposed via <sys/prctl.h> or <linux/prctl.h>.
 #include <asm/prctl.h>
 // This syscall stub is generated but it's not declared in any header.
@@ -104,4 +83,6 @@ static inline int __set_tls(void* tls) {
 #error unsupported architecture
 #endif
 
+#include "tls_internal.h"
+
 #include "tls_defines.h"
diff --git a/libc/platform/bionic/tls_internal.h b/libc/platform/bionic/tls_internal.h
new file mode 100644
index 000000000..cadfd2c9d
--- /dev/null
+++ b/libc/platform/bionic/tls_internal.h
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2008 The Android Open Source Project
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *  * Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef __BIONIC_PRIVATE_GET_TLS_INTERNAL_H_
+#define __BIONIC_PRIVATE_GET_TLS_INTERNAL_H_
+
+#if defined(__aarch64__)
+# define __get_tls_internal() ({ void** __val; __asm__("mrs %0, tpidr_el0" : "=r"(__val)); __val; })
+#elif defined(__arm__)
+# define __get_tls_internal() ({ void** __val; __asm__("mrc p15, 0, %0, c13, c0, 3" : "=r"(__val)); __val; })
+#elif defined(__i386__)
+# define __get_tls_internal() ({ void** __val; __asm__("movl %%gs:0, %0" : "=r"(__val)); __val; })
+#elif defined(__riscv)
+# define __get_tls_internal() ({ void** __val; __asm__("mv %0, tp" : "=r"(__val)); __val; })
+#elif defined(__x86_64__)
+# define __get_tls_internal() ({ void** __val; __asm__("mov %%fs:0, %0" : "=r"(__val)); __val; })
+#else
+#error unsupported architecture
+#endif
+
+#endif /* __BIONIC_PRIVATE_GET_TLS_INTERNAL_H_ */
diff --git a/libc/platform/scudo_platform_tls_slot.h b/libc/platform/scudo_platform_tls_slot.h
index a86b828f2..f668e83b9 100644
--- a/libc/platform/scudo_platform_tls_slot.h
+++ b/libc/platform/scudo_platform_tls_slot.h
@@ -31,5 +31,5 @@
 #include "bionic/tls.h"
 
 inline uintptr_t* _Nonnull getPlatformAllocatorTlsSlot() {
-  return reinterpret_cast<uintptr_t*>(&__get_tls()[TLS_SLOT_SANITIZER]);
+  return reinterpret_cast<uintptr_t*>(&__get_tls_hooks()[TLS_SLOT_SANITIZER]);
 }
